### ISSUE TITLE
Multiple encodings, much cores, low CPU utilization

### ISSUE THREAD

**Issue by MYodaTheRe** [RandomEngy/VidCoder#478]
Hi,

i have found another issue in 4.28 / 4.29+

First, I have a workstation with dual socket: 2x Xeon X5650.
This are 2x6 "real" cores, so no hyperthreading is enabled. 6 Cores per socket.

If I enable multiple encodings in the global settings and configure it two use 2 encodings in parallel only the CPU power of 1 socket will be used. So CPU utilization of the whole machine will be on about 45-50% during 2 encodings in parallel.

If I do 3 multiple encodings, every process slows down to about 15%~ CPU usage (instead of 25%~ while 2 encodings); therefore again only 45-50% of the CPU resources will be used.

Is there a possibility that you could add a setting for CPU / socket bindings of encodings or fix this behaviour another way?

I use:
- x265 (8bit)
- 1080p (so no UHD)
- default settings with additional "--frame-threads=1", because i want maximum possible qualitiy aswell as little files and the x265 docs say, that every more frame-thread will eventually reduce quality.

**Comment by RandomEngy** [comment]
As long as CPU throttling is set to 100% of allowed cores under Global options -> Process, it shouldn't be constraining CPU use to a specific socket.

You might have another bottleneck in the system like disk read/write speed.

**Comment by MYodaTheRe** [comment]
CPU throttling is at 100%, worker priority is to below normal.
It is quite old hardware (HP Z600 Workstation, Xeon CPU release was 2010), but the newest Windows 10 with all patches is installed.

RAM usage is 3,6GB / 12 GB, the disc cannot be the issue, since I use a dedicated NAS to read the source and write the rendered data to. This NAS is capable of doing 2Gbit/s without problems.

I have linked a screenshot of my taskmanager here, you can clearly see, that only one socket is used. The only thing I can do is, to change the socket membership manually in the task manager, but it will be resetted after each sucessfull encode:

[https://imgur.com/a/9WK8qLl](url)

**Comment by MYodaTheRe** [comment]
I found the cause of the issue: I throttled the maximum CPU usage in the Windows 10 energy saving plan to 95% (maximum cpu usage), because of overheating problems in the past.

If I set this to 100% aswell, both sockets will be used. I don't know why only one socket will be used, if I set the limit to 95%, because it should only throttle the GHz of the CPU cores to 95% of the maximum.
Does this error associate specifically with VidCoder or is this a Windows 10 Problem in general?

**Comment by RandomEngy** [comment]
Sounds like a Windows issue then. VidCoder doesn't do anything with processor sockets.

**Comment by MYodaTheRe** [comment]
> 
> 
> I found the cause of the issue: I throttled the maximum CPU usage in the Windows 10 energy saving plan to 95% (maximum cpu usage), because of overheating problems in the past.
> 
> If I set this to 100% aswell, both sockets will be used. I don't know why only one socket will be used, if I set the limit to 95%, because it should only throttle the GHz of the CPU cores to 95% of the maximum.
> Does this error associate specifically with VidCoder or is this a Windows 10 Problem in general?

Well I have to correct this, as soon as the encoding was finished and the next started, only one of the two sockets was used again.
Would it be possible for you to add another option under Global -> process (maybe a checkbox). This option could schedule new worker processesses to sockets like:
process 1 -> socket 1 (if there is no current other worker process binding to this socket AND there is another "free" socket)
process 2 -> socket 2 (if there is no current other worker process binding to this socket AND there is another "free" socket)

...
and so on

This ensures, that you can start more than two worker processesses on two sockets, because then the first socket would get the third worker process.

**Comment by MYodaTheRe** [comment]
Yeah, i changed it through the "set affinity" option, by default (and after/for every new vidcoder worker instance) there are selected any CPUs, so CPU 0 to CPU 5 (Node 0) and CPU 0 to CPU 5 (Node 1).

However both worker processes are stuck to one node/socket, until i manually deactivate CPUs (for example all CPUs from Node 0) for one of the worker processes.

**Comment by MYodaTheRe** [comment]
> 
> 
> Sorry, that wasn't clear, are you saying that it had affinity checked of all cores by default? 

Yeah, every time a job is finished and the next worker process starts, all cores will be enabled by default - but only half the total cpu power will be used by all vidcoder instances with all cpu cores activated.

>And you could only change it by deactivating CPUs? This really sounds like a Windows/Hardware problem.
> 
> I found someone had a similar issue that was fixed by a BIOS update: https://stackoverflow.com/questions/28098082/unable-to-use-more-than-one-processor-group-for-my-threads-in-a-c-sharp-app

Hmm, interesting since I have a HP machine, too.
I will check out for maybe available BIOS update.

**Comment by MYodaTheRe** [comment]
I have installed the latest BIOS update now, the update itself didn't change anything.

BUT I could enable a BIOS setting (chipset/storage) -> storage node interleave
This was disabled before and as I enabled it, it fixed the socket problem.

Now all CPUs are used without manual affinity setting.
